text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,165 @@
+
+The semantic conventions described below define how to model those stages in
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+Two layers of context propagation need to be distinguished for messaging
+workflows:
+
+1. The _creation context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. The _transport context layer_ allows correlating the producer and the
+   consumer with an intermediary. It also allows to correlate multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries
+   that are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+A producer SHOULD attach a creation context to each message. The creation context
+SHOULD be attached in a way so that it is not changed by intermediaries. A
+producer MAY propagate a transport context to an intermediary.  An
+intermediary MAY propagate a transport context to a consumer.
+
+### Trace structure, names, and attributes
+
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Deliver"
+spans to correlate producer with consumer traces. A "Deliver" span must link
+to the "Create" spans of all messages that are handled by the respective
+"Deliver" operation. Depending on the use case, "Deliver" spans can correlate
+with "Process" spans or other spans modelling processing operations.
+
+The operation modelled by the "Deliver" span does not strictly refer to
+receiving the message from intermediaries, but instead refers to the
+application receiving messages for processing. If messages are fetched from the
```

Comment:
An open point from today's workgroup discussion:

The duration of a "Deliver" span as specified here might involve the duration of "Process" spans (e. g. when the "Deliver" span is created for a callback invocation and messages are processed in that callback). However, for some scenarios the duration of the "Deliver" span does not involve the duration of "Process" spans (e. g. when the "Deliver" span is just created for a call that passes messages to the application for processing, and "Process" spans are created after the "Deliver" operation ends).

Some questions need to be answered:
* Is it a problem that the "Deliver" span can have those different semantic meanings?
* Should we distinguish both scenarios? If yes, how? Span names, or attributes?

@lmolkova @blumamir

Reply:
> Is it a problem that the "Deliver" span can have those different semantic meanings?

I believe yes - from user (e,g, Ops or new engineer on the team) perspective if I just look at spans without enough knowledge of the codebase, I should be able to tell if this span (duration, outcome) covers the whole processing or just delivery. So these two have different semantics. 

> Should we distinguish both scenarios? If yes, how? Span names, or attributes?

I propose to start with different names (as we didn't yet start to discuss attributes).
`Delivery` seems to precisely describe passing events to the application, but only it.  'Process' is very precise in describing what happens in the callback.

Reply:
Another way (i believe on top of naming) is to have different kinds: 
- Deliver that involves delivery only is CLIENT or INTERNAL span
- Processing is a CONSUMER span 

BTW looking into [SpanKind](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind) definitions - we'll need to sync them along with messaging stabilization

Reply:
This concern should be covered by having different span names for pull ("Receive") and push ("Deliver") based scenarios.